File: src/main/ruby/advent2023/Puzzle1.rb
load 'base.rb'

module Puzzle1
  class Puzzle
    include Base

    def main
      solution1 = self.solve_puzzle("puzzle1/inputs.txt", Array.new) do |line, data|
        data << num_from_str(line)
      end
      puts "Solution1: #{solution1}"

      solution2 = self.solve_puzzle2("puzzle1/inputs.txt", Array.new) do |line, data|
        data << num_from_str2(line)
      end
      puts "Solution2: #{solution2}"
    end

    def compute_solution(data)
      data.sum
    end

    def compute_solution2(data)
      data.sum
    end

    def num_from_str(str)
      regex = /(\d)/
      first_int = str[regex].to_i
      second_int = str[str.rindex(regex)].to_i
      (first_int * 10) + second_int
    end

    def get_all_nums(str)
      ret = []
      str_nums = /(one|two|three|four|five|six|seven|eight|nine)/
      map = {"one"=>1,"two"=>2,"three"=>3,"four"=>4,
        "five"=>5,"six"=>6,"seven"=>7,"eight"=>8,"nine"=>9}
      (0..str.length).each do |i|
        substr = str[i..]
        if substr.start_with? str_nums
          ret << map[substr.match(str_nums)[0]]
        elsif substr.start_with? /\d/
          ret << str[i].to_s.to_i
        end
      end
      ret
    end

    def num_from_str2(str)
      scans = get_all_nums(str)
      first_int = scans[0]
      second_int = scans[-1]
      (first_int * 10) + second_int
    end
  end
end

p = Puzzle1::Puzzle.new
p.main

File: src/main/python/advent2024/Puzzle1.py
from Base import Base
from typing import Tuple

Entry = Tuple[int, int]
MyDataType = list[Entry]

class Puzzle(Base[MyDataType, int, int]):
    def main(self):
      solution1: int = self.solve_puzzle("puzzle1/inputs.txt", [])  # type: ignore
      print("Solution1: {}".format(solution1))

      solution2: int = self.solve_puzzle2("puzzle1/inputs.txt", []) # type: ignore
      print("Solution2: {}".format(solution2))

    def read_one_line(self, line: str, data: MyDataType):
      split = line.split()
      entry = int(split[0]), int(split[1])
      data.append(entry)

    def compute_solution(self, data: MyDataType) -> int:
      sorted_first = sorted(map(lambda x: x[0], data))
      sorted_second = sorted(map(lambda x: x[1], data))
      diff = 0
      for i in range(len(sorted_first)):
          diff += abs(sorted_first[i] - sorted_second[i])
      return diff

    def compute_solution2(self, data: MyDataType):
      mapped_second = {}
      for element in data:
          key = element[1]
          if key not in mapped_second:
              mapped_second[key] = 0
          mapped_second[key] += 1
      answer = 0
      for element in data:
          key = element[0]
          if key in mapped_second:
              answer += mapped_second[key] * key
      return answer
    
puzzle: Puzzle = Puzzle()
puzzle.main()